daemon: Add --kube-proxy-replacement flag

brb opened the pull request:
This PR:
- Introduces the agent's `--kube-proxy-replacement` flag. The flag controls what / how datapath features required for the kube-proxy replacement enabled. The possible outcomes depending on a value of the flag are the following:
    - "probe": auto-enable all available features for kube-proxy replacement (agent probes for features and tries to enable them without panicking if any cannot be enabled).
    - "strict": enable all features (panic if any cannot be enabled).
    - "partial": enable only selected by a user features (panic if any of the selected ones cannot be enabled).
    - "disabled": disable all kube-proxy replacement features (even those which user has enabled).

- Introduces the `kubeProxyReplacement` helm var (defaults to `"probe"`).
- Adds a probe for the NodePort feature.

The upgrade considerations:

- New users will have the kube-proxy replacement auto-enabled (`"probe"`) due to the helm var default value.
- Existing users are encouraged via the upgrade guide to set the option to `"strict"` if previously they were running with `enable-node-port=true`

Reviewable per commit.

The docs update and showing which features are enabled from `cilium status` will be added as a follow-up (once we finally decide on the flag name and its semantics).

```release-note
Add --kube-proxy-replacement flag to control enabling of kube-proxy replacement in BPF
```

Fixes #9729

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9992)
<!-- Reviewable:end -->

brb committed:
agent: Rename --enable-{k8s-,}external-ips flag

In the datapath we call the feature as "ExternalIPs", as it is
independent of Kubernetes. Therefore, rename the flag to reflect that,
but keep "k8s" in the help description, so that users would find it (I
doubt that anyone will want to mess with this flag).

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Move validation of svc related flags into function

For the sake of readability.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
probes: Probe for BPF helpers

This commit adds GetHelpers() method which returns available helper
functions for the given program type.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm commented on daemon/daemon_main.go:
```diff
@@ -481,6 +482,9 @@ func init() {
 	flags.StringSlice(option.Labels, []string{}, "List of label prefixes used to determine identity of an endpoint")
 	option.BindEnv(option.Labels)
 
+	flags.Bool(option.EnableKubeProxyFree, false, "Enable kube-proxy replacement in Cilium datapath (BPF)")
```
Maybe add something "Enable kube.... (eBPF) if supported"

borkmann replied:
Maybe: `Enable or partially enable kube-proxy BPF replacement in Cilium datapath if supported by the kernel`

borkmann commented on daemon/daemon_main.go:
```diff
@@ -1410,7 +1411,16 @@ func (d *Daemon) instantiateAPI() *restapi.CiliumAPI {
 func initKubeProxyFreeOptions() {
-	if option.Config.EnableNodePort &&
+	if option.Config.EnableKubeProxyFree {
+		option.Config.EnableNodePort = true
```
Should we also include an override of`option.Config.EnableExternalIPs` to true here?

borkmann commented on daemon/daemon_main.go:
```diff
@@ -1421,32 +1431,56 @@ func initKubeProxyFreeOptions() {
+			} else {
+				log.Warn("BPF host reachable services for UDP needs kernel 4.19.57, 5.1.16, 5.2.0 or newer. Disabling the feature.")
+				option.Config.EnableHostServicesUDP = false
+			}
```
If both are false we should also set `option.Config.EnableHostReachableServices` to false.

borkmann commented on daemon/daemon_main.go:
```diff
@@ -1421,32 +1431,56 @@ func initKubeProxyFreeOptions() {
+				log.Fatal("BPF NodePort services needs kernel 4.17.0 or newer")
+			} else {
+				log.Warn("BPF NodePort services needs kernel 4.17.0 or newer. Disabling the feature.")
+				option.Config.EnableNodePort = false
```
And here we should also set `option.Config.EnableExternalIPs` to false.

borkmann replied:
Aside form the few nits and remaining TODOs, looks great!

brb committed:
daemon: Add --kube-proxy-replacement flag

The flag controls what / how datapath features required for kube-proxy
replacement enabled. The possible outcomes depending on a value of the
flag are the following:

- "probe": auto-enable all available features for kube-proxy replacement
  (agent probes for features and tries to enable them without panicking
   if any cannot be enabled).
- "strict": enable all features (panic if any cannot be enabled).
- "partial": enable only selected by a user features (panic if any of
  the selected ones cannot be enabled).
- "disabled": disable all kube-proxy replacement features (even those
  which user has enabled).

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
helm: Add kubeProxyReplacement

It controls the agent's --kube-proxy-replacement flag.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

joestringer commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -113,6 +113,7 @@ cilium-agent [flags]
       --k8s-watcher-queue-size uint                           Queue size used to serialize each k8s event type (default 1024)
       --keep-bpf-templates                                    Do not restore BPF template files from binary
       --keep-config                                           When restoring state, keeps containers' configuration in place
+      --kube-proxy-replacement string                         Detect and enable individual features for kube-proxy replacement ("strict"), or enable only selected features and panic if dependencies between features are broken ("strict"), or enable only selected features regardless dependencies ("partial")  (default "strict")
```
```suggestion
      --kube-proxy-replacement string                         Detect and enable individual features for kube-proxy replacement ("probe"), or enable only selected features and panic if dependencies between features are broken ("strict"), or enable only selected features regardless dependencies ("partial")  (default "strict")
```

joestringer replied:
Is there any reason to do "partial"? What does that enable users to do? They'll still need kube-proxy.

brb replied:
The main idea of "partial" is to run together with kube-proxy and to offload only some its functionality to cilium.

E.g. a user wants to enable NodePort for outside connections (handled by bpf_netdev), but at the same time they want to disable the host-reachable services (e.g. due to https://github.com/cilium/cilium/issues/9974) and instead to use kube-proxy for accesing NodePort from inside cluster.

joestringer commented on pkg/option/config.go:
```diff
@@ -755,6 +760,18 @@ const (
+
+	// KubeProxyReplacementStrictPartial specifies to strictly enable selected
+	// kube-proxy replacement features
+	KubeProxyReplacementStrictPartial = "strict"
```
```suggestion
	KubeProxyReplacementStrict = "strict"
```

joestringer commented on daemon/daemon_main.go:
```diff
@@ -481,6 +482,13 @@ func init() {
+		"Detect and enable individual features for kube-proxy replacement (%q), "+
+			"or enable only selected features and panic if dependencies between features are broken (%q), "+
+			"or enable only selected features regardless dependencies (%q) ",
+		option.KubeProxyReplacementStrictPartial, option.KubeProxyReplacementStrictPartial, option.KubeProxyReplacementPartial))
```
```suggestion
		option.KubeProxyReplacementProbe, option.KubeProxyReplacementStrict, option.KubeProxyReplacementPartial))
```

joestringer reviewed (changes requested):
I primarily focused on docs / datapath piece, didn't look in detail at daemon / CLI. Happy to do so if you don't get such review from @cilium/agent .

One other comment on upgrade above.

brb commented on daemon/daemon_main.go:
```diff
@@ -1454,3 +1413,111 @@ func (d *Daemon) instantiateAPI() *restapi.CiliumAPI {
+	}
+
+	if option.Config.EnableNodePort && option.Config.EnableIPSec {
+		log.Fatal("IPSec cannot be used with NodePort BPF")
```
Still not sure about this one. Maybe in a non-strict mode we should just log a warning, disable the NP feature, and proceed further? I'm afraid that users who have the IPSec enabled, and haven't closely followed the upgrade guide and have re-generated helm charts during the upgrade will end up with cilium panicking.

brb commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -113,6 +113,7 @@ cilium-agent [flags]
       --k8s-watcher-queue-size uint                           Queue size used to serialize each k8s event type (default 1024)
       --keep-bpf-templates                                    Do not restore BPF template files from binary
       --keep-config                                           When restoring state, keeps containers' configuration in place
+      --kube-proxy-replacement string                         Detect and enable individual features for kube-proxy replacement ("strict"), or enable only selected features and panic if dependencies between features are broken ("strict"), or enable only selected features regardless dependencies ("partial")  (default "strict")
```
Shall we add "disabled" as well? It's not so straightforward to disable it ("strict" + each feature has to be disabled).

aanm replied:
yeah, a disable here would be more explicit.

brb committed:
docs: Mention kube-proxy-replacement in upgrade docs

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -113,6 +113,7 @@ cilium-agent [flags]
       --k8s-watcher-queue-size uint                           Queue size used to serialize each k8s event type (default 1024)
       --keep-bpf-templates                                    Do not restore BPF template files from binary
       --keep-config                                           When restoring state, keeps containers' configuration in place
+      --kube-proxy-replacement string                         Detect and enable individual features for kube-proxy replacement ("strict"), or enable only selected features and panic if dependencies between features are broken ("strict"), or enable only selected features regardless dependencies ("partial")  (default "strict")
```
s/`option.KubeProxyReplacementStrictPartial, option.KubeProxyReplacementStrictPartial, `/`KubeProxyReplacementProbe, option.KubeProxyReplacementStrictPartial`

aanm commented on daemon/daemon_main.go:
```diff
@@ -1407,7 +1415,38 @@ func (d *Daemon) instantiateAPI() *restapi.CiliumAPI {
+		option.Config.EnableHostServicesUDP = true
+	}
+
+	if option.Config.EnableNodePort && option.Config.EnableIPSec {
```
I think a `switch case` would make this a little more readable .
```
switch option.Config.KubeProxyReplacement {
   case option.KubeProxyReplacementProbe:
    ...
    default:
    log.Fatalf...
}
```

brb replied:
Wouldn't that lead to a lot of code duplication?

aanm commented on daemon/daemon_main.go:
```diff
@@ -1407,7 +1415,38 @@ func (d *Daemon) instantiateAPI() *restapi.CiliumAPI {
+
+		if option.Config.NodePortMode == "dsr" && option.Config.Tunnel != option.TunnelDisabled {
+			log.Fatal("DSR cannot be used with tunnel")
+		}
```
Please specify the "why" they are invalid in code comments.

aanm commented on pkg/option/config.go:
```diff
@@ -755,6 +760,18 @@ const (
+
+	// KubeProxyReplacementPartial specifies to enable only selected kube-proxy
+	// replacement features
+	KubeProxyReplacementPartial = "partial"
```
```
type KubeProxyReplacementMode string

const (
        KubeProxyReplacementProbe KubeProxyReplacementMode = "probe"
        KubeProxyReplacementStrict  KubeProxyReplacementMode = "strict"
        KubeProxyReplacementPartial KubeProxyReplacementMode = "partial"
)
```

brb replied:
Not sure whether an extra type would give us any benefits besides adding more code. Also, it's not consistent with existing flags, e.g. `--tunnel`.

aanm commented on install/kubernetes/quick-install.yaml:
```diff
@@ -121,6 +121,7 @@ data:
 
   install-iptables-rules: "true"
   auto-direct-node-routes: "false"
+  kube-proxy-replacement:  "probe"
```
Shouldn't all other options be enabled by default in helm? `external-ips` `node-port`, `host-rechable-services`

brb replied:
I don't have a strong opinion on this. If we default `kube-proxy-replacement` to `probe`, then majority of users won't need to mess with those options anyway. If I user sets the option to `strict` or `partial`, then we assume that they know what they are doing, so they can set manually each individual option.

brb committed:
test: Set kubeProxyReplacement=strict when running w/o kube-proxy

To avoid cilium-agent silently not enabling all features, which we want
to test when running w/o kube-proxy.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -406,6 +406,18 @@ New ConfigMap Options
+    enable only those features which user has explicitly enabled in their
+    ConfigMap. See :ref:`kubeproxy-free` for more option values.
+
+    For users who previously were running with ``nodePort.enables=true`` it is
```
```suggestion
    For users who previously were running with ``nodePort.enabled=true`` it is
```

joestringer commented on daemon/daemon_main.go:
```diff
@@ -1454,3 +1415,130 @@ func (d *Daemon) instantiateAPI() *restapi.CiliumAPI {
+			}
+		}
+		log.WithField(logfields.Interface, device).
+			Info("Using auto-derived device for BPF node port")
```
I think that this message will always be printed even if the feature was disabled due to a non-fatal error a few lines up, in `--probe` mode?

brb replied:
Nice spot!

joestringer reviewed (changes requested):
Couple of minor nits below, otherwise LGTM.
